### PR TITLE
First draft of the SPL, UnitDSL and NESTML parts

### PR DESCRIPTION
**Pull request by DimitriPlotnikov**

### CODE REVIEW

**Comment by jougs** [comment]
@DimitriPlotnikov: I've just merged #4 by @ingablundell, which leads to conflicts with your PR. Can you please run `git pull jougs reference_documentation` in your branch to update the PR and push the result?

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: In this context I'd rather keep it as open as possible. What about "programming, description and markup languages"?

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: I know that I wrote this. What do you think about replacing "connection" by "forced correspondence"?

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Move "freedom" after "complete" earlier in the sentence.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Is this true? I thought `integer` corresponds to `long`.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "present" by "lists the"

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Link "Le Système internationale d’unités" to https://en.wikipedia.org/wiki/International_System_of_Units.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "support convenience in" by "ease".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Enclose all operators in back ticks.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:91]: Move "Also" behind "can".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Move "also" in front of "checked".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "Further" by "Furthermore".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Add "an" in front of "arbitrary".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Move "Also" behind "must".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "to" by "with".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Use single instead of triple back ticks in inline code.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Make "Functions" lower case.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:158]: Should we add some more comprehensive explanations here? Some of these have already been explained by @ingablundell in #4. Maybe references to the corresponding sections are enough?

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:226]: Add one explanatory sentence per example.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:301]: Also add an explanatory sentence per example here.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Put variable names in separate back ticks separated by plain text commas.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Put operators and expressions in back ticks.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:37]: Isn't the part about "zero or more arguments" true? Cf. for example `neuron <name>`

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "with" by "of".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Remove "concrete".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Remove first occurrence of "neuron".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "with" by "of".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Remove superfluous space.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "terminated" by "is closed".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "with" by "of".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:366]: I don't think that these should be constant, as this defies the purpose. However, if the current implementation is like this, we can leave it unchanged until fixed.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "details later" by "detail later on".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "details later" by "detail later on".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "kinds of events" by "type of event" and "fire" by "send"

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace first occurrence of "the" by "this".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:388]: Also explain that this block is the central piece of code for updating the dynamics of the neuron and that the pre-defined variable `t` is available here and that the block is executed in the update loop of NEST.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "an" by "a" and "round brackets" by "parentheses".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace first occurence of "the" by "this".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:403]: Add reference to the function documentation above. Anchors are available for each (sub-) section under its name in all lower case letters and spaces replaced by dashes. I.e. for a section
```
### The neuron dynamics
```
a link can be generated using
```
[section about the neuron model](#the-neuron-model)
```

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Remove first occurrence of "as well" and replace "positive" by "positively" and "negative" by "negatively".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "positive" by "positively" and "negative" by "negatively".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "are" by "is" and first occurrence of "the" by "than".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Add "status" in front of "dict".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: I seem to remember that we once discussed replacing "emit_spike()" by just "emit()" as the type of output event can in principle be also something else than spike. However, if the current implementation is like this, we can leave it unchanged until fixed.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:492]: Add empty line above heading.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:493]: Replace "refractory" by "refractoriness" and add "the" in front of "non.refractory".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md:494]: Add "The" in front of "First" and replace "refractoriness period" by "refractory time".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Replace "delay" by "passed time span".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Add "using" behind "read".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Add "the keyword" in front of "recordable".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Remove the final sentence.

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Remove second occurrence of "in".

File: doc/syntax.md
Comment (jougs) [inline doc/syntax.md]: Add "the" behind "during".

**Review by jougs — CHANGES_REQUESTED** [review]
Many thanks for this extensive and solid work! My comments are mainly minor corrections and fixes to spelling and grammar.

The PR should apply cleanly once you merge in my current state (see previous comment).

I'll approve as soon as my concerns are addressed.

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md]: Is OK for me. IMHO: it is an important selling point of nestml that it is a DSL

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md]: Yes, you are right. Do we need an additional type? We could even make the type mapping flexible and adjustable by the user.

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md]: ok, also examples are put into back ticks

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md]: Complementary, I will extend my tutorial (not reference manual) with a list of context conditions (about 30). After that we can copy and paste them into manual. They, actually define the type system of nestml.

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md:158]: Yes, we should. But in the next iteration. Also number and type of arguments should be listed.

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md:37]: Yes, `state` has zero arguments. `neuron`, `function` one or more "arguments".

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md:366]: They are constant. I have good reasons to do them constant. If you do want mutable variables, define them in directly in the update block. Otherwise, I cannot do some constant folding optimizations and NESTML code will not meet performance of the handwritten code.

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md:388]: Yes, but for NOW `t` is cannot be accessed in the dynamics block. It will be possible in the next nestml release (I have to add a bunch of context conditions to restrict the usage of t to a) equations b) update). In the remaining block it makes no sense to use allow `t`

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md:403]: I don't get the point. Here I introduce the syntax for functions and provide an example. Which place in text should I reference?

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md]: spikes are routed -> spikes IS routed?
to the `excitatory` port -> to THAN excitatory` port?
It sounds wrong for me.

**Comment by DimitriPlotnikov** [comment]
@jougs I worked trough your findings and commented most of your concerns. For mysterious reasons, github doesn't render tables properly. Could you take look at it? I'm working with atom and it shows them correctly.

File: doc/syntax.md
Comment (jougs) [reply doc/syntax.md]: We could also point this out in the first sentence: "NESTML is a domain specific language which has a concise..."

File: doc/syntax.md
Comment (jougs) [reply doc/syntax.md]: If we introduce an additional type, we should call it something like integer32 to indicate its size. But right now, I don't think we need it. Configurability might be needed if we start to support other backends than NEST. Right now, I think things can stay as they are.

File: doc/syntax.md
Comment (jougs) [reply doc/syntax.md]: Perfect!

File: doc/syntax.md
Comment (jougs) [reply doc/syntax.md:158]: Fair enough.

File: doc/syntax.md
Comment (jougs) [reply doc/syntax.md:37]: So why did you remove it from the text?

File: doc/syntax.md
Comment (jougs) [reply doc/syntax.md:403]: To somewhere around line 150, where the `function` block is explained.

File: doc/syntax.md
Comment (jougs) [reply doc/syntax.md]: Yes, that would totally be wrong. The comment refers to the line directly above it:
```
There IS more THAN one `spike` or `current` port.
```

**Comment by jougs** [comment]
Many thanks for answering my comments. I've again replied where required.

Regarding the tables: https://help.github.com/articles/organizing-information-with-tables/ has the official documentation for tables in GitHub Flavored Markdown. I usually use the `kramdown` engine with the `-i GFM` option to render locally.

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md:403]: Function calls are explained there. Still don't understand, why should reference there. If you want, just add the reference.

File: doc/syntax.md
Comment (DimitriPlotnikov) [reply doc/syntax.md:37]: It is some pr/merge magic. Comparing the resulting text, I'll find this paragraph there.

**Comment by jougs** [comment]
Thanks for answering. I'll make another short round through the current version of the text and will merge immediately if no major changes are required.

**Comment by DimitriPlotnikov** [comment]
Ready for review. 

Tool you proposed doesn't run on my operating system. A docker container would be nice.
There is one issue in the table: List of operators. The bitwise or operator(`|`) is currently added as an unicode code. Otherwise, it is interpreted as column separator. Can you fix it?

**Review by jougs — APPROVED** [review]
Thanks! I think I'll take it from here and try to fix typos and structure before I start the pull request against upstream master.

**Pull request merged**
